### ISSUE TITLE
Use the AtLeastOne generic type to enforce field counts

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#2028]
**Scheduling**

This risk is acceptable for the stable release, but we need to fix it before we support lightwalletd.

**Is your feature request related to a problem? Please describe.**

Currently, Zebra has 3 copies of similar code, to make sure that there is at least one JoinSplit, and sapling Spend, and sapling Output in particular structs or enum variants.

We could combine this code into a single generic `AtLeastOne` struct.

**Describe the solution you'd like**

Stage 1:

Implement `AtLeastOne`:
- [x] Add a generic `AtLeastOne<T>` wrapper for `Vec<T>`, with:
  - a `TryFrom<Vec<T>, SerializationError>` impl
  - a `Deref<Target = Vec<T>>` impl
    - this impl gets us `impl Iterator` and `Deref<[T]>` for free, and all the related traits (like `DoubleEndedIterator`)
  - ZcashSerialize and ZcashDeserialize impls
  - a `first(&self) -> &T` method, to simplify the block coinbase check
  - a `split_first(&self) -> (&T, &[T])` method, to simplify working with non-coinbase transactions
  - derive `Eq`, `PartialEq`, `Ord`, `PartialOrd`, `Clone`, `Debug`, `Serialize`, `Deserialize`

Use `AtLeastOne`:
- [x] Replace the first and rest fields in `TransferData` in #2021 with `AtLeastOne` (for both `Spend` and `Output` `first_*` and `rest_*`, but not `maybe_outputs`)

Stage 2:
- [ ] Use `AtLeastOne` in `JoinSplitData`
  - we might need a generic serde helper here, or we can just serialise using the iterator or `Vec`

Stage 3:
- [ ] Use `AtLeastOne` in `Block.transactions`
- [ ] Remove any runtime checks that there is at least one transaction, they are now enforced during deserialisation

Stage 4:
- [ ] Use `AtLeastOne` in external `Message` variants `GetBlocks` and `GetHeaders`
  - other variants with arrays allow zero entries 😞 
- [ ] Remove any runtime checks, they are now enforced during deserialisation

Stage 5:
- [ ] Use `AtLeastOne` for all arrays in internal `Request`s and `Response`s
  - drop invalid `Message`s during conversion, rather than errorring - they are valid externally, but we don't want to have to special-case them internally
  - check that response processing code doesn't depend on any messages we've dropped
- [ ] Remove any runtime checks, they are now enforced during deserialisation

**Describe alternatives you've considered**

We could keep things as they are, and continue to use duplicate code and runtime checks across the codebase.

**Additional context**

Stage 1 depends on #2021 and stage 2 has tricky serde bounds. So we might actually want to do Stage 3 first.

We might want to do Stage 1 as part of #2021, so we can simplify the design and code review.

This cleanup task will probably be completed over a few sprints. Stages 4 and 5 are low priority, because network messages aren't part of the consensus rules.

**Comment by teor2345** [comment]
This is a useful validation, but it's not a high priority right now.
